### ISSUE TITLE
Dex OAUTH: "invalid client credentials"

### ISSUE THREAD

**Issue by elouanKeryell-Even** [argoproj/argo-cd#17088]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [ ] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

Authentication through UI often fails with the following error:

```
failed to get token: oauth2: "invalid client" "invalid client credentials."
```

(see screenshots further down)

sometimes, opening a new tab, re-typing argocd homepage URL and retrying the login process, makes it work

Our setup:
- two argocd-server
- a dex server
- a dex OAUTH connector pointing to our company OAUTH server
- a "sticky session" setup using istio

**IMPORTANT**: something we noticed is the errors seem to occur when interacting with the OLDEST INSTANCE

*our two argocd-server instance:*

```
$ kubectl -n argocd-system get pods -l app.kubernetes.io/name=argocd-server
NAME                             READY   STATUS    RESTARTS   AGE
argocd-server-79bc4c4f57-4g5sd   2/2     Running   0          3d
argocd-server-79bc4c4f57-njhwx   2/2     Running   0          12d
```

*dex config (in configmap `argocd-cm`)*:

```yaml
connectors:
  - type: oauth
    id: sesame
    name: sesame
    config:
      authorizationURL: https://***:443/***/oauth2/multiauth/authorize?acr_values=sesameGROUPE

      claimMapping:
        preferredUsernameKey: sub
        userNameKey: sub
      
      # Inject sensitive values from secret `argocd-secret`
      # See here: https://argo-cd.readthedocs.io/en/stable/operator-manual/user-management/#sensitive-data-and-sso-client-secrets
      clientID: ***
      clientSecret: $dex.config.sesame.clientSecret

      redirectURI: https://argocd.hp.arom.dco.aws.internal.cloud.***/api/dex/callback
      scopes:
        - email
        - openid
        - uid
      tokenURL: https://***:443/***/oauth2/multiauth/access_token
      userIDKey: sub
      userInfoURL: https://***:443/***/oauth2/multiauth/userinfo
```

*istio destination rule to setup sticky session for our multiple argocd-server instances*:

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: argocd-sticky-session
  namespace: argocd-system
spec:
  host: argocd-server.argocd-system.svc.cluster.local
  trafficPolicy:
    loadBalancer:
      consistentHash:
        httpCookie:
          name: user
          ttl: 36000s
```

**To Reproduce**

Error does not happen all the time, but here is my process:

- open argocd
- chose to authenticate through OIDC
- get redirected to OIDC provider
- log in successfully
- get error message on callback to argocd

**Expected behavior**

I expect to be able to successfully login

**Screenshots**

![InkedCapture2_LI](https://github.com/argoproj/argo-cd/assets/9483289/c4caf179-ba8d-466f-bdb9-a07873820765)

![Capture3-2](https://github.com/argoproj/argo-cd/assets/9483289/868bea66-98d1-4a1a-98f9-33d6fa9cbf1c)

![InkedCapture_LI](https://github.com/argoproj/argo-cd/assets/9483289/8c5b746a-0e7f-414a-8d68-9b3ae33edc6c)

*When a login works, it looks like this:*

![image](https://github.com/argoproj/argo-cd/assets/9483289/3c1b5b05-b98d-40a1-9606-51554fc83d53)

**Version**

```shell
$ kubectl -n argocd-system exec -it deployment/argocd-server -- argocd version
argocd: v2.8.4+c279299
  BuildDate: 2023-09-13T19:12:09Z
  GitCommit: c27929928104dc37b937764baf65f38b78930e59
  GitTreeState: clean
  GoVersion: go1.20.6
  Compiler: gc
  Platform: linux/amd64
FATA[0000] Argo CD server address unspecified
command terminated with exit code 1
```

**Logs**

Login failure:

```
### argocd server
argocd-server-79bc4c4f57-njhwx server time="2024-02-05T09:14:01Z" level=info msg="Performing authorization_code flow login: https://argocd.hp.arom.dco.aws.internal.cloud.***/api/dex/auth?client_id=argo-cd&redirect_uri=https%3A%2F%2Fargocd.hp.arom.dco.aws.internal.cloud.***%2Fauth%2Fcallback&response_type=code&scope=openid+profile+email+groups&state=***"
argocd-server-79bc4c4f57-njhwx server time="2024-02-05T09:14:11Z" level=info msg="Callback: /auth/callback?code=***&state=***"

### dex server
time="2024-02-05T09:14:11Z" level=info msg="login successful: connector \"sesame\", username=\"***\", preferred_username=\"***\", email=\"***@*** (unverified)\", groups=[]"
time="2024-02-05T09:14:11Z" level=info msg="invalid client_secret on token request for client: argo-cd"
```

Login success:

```
### argocd server
argocd-server-79bc4c4f57-njhwx server time="2024-02-05T09:39:08Z" level=info msg="Performing authorization_code flow login: https://argocd.hp.arom.dco.aws.internal.cloud.***.***/api/dex/auth?client_id=argo-cd&redirect_uri=https%3A%2F%2Fargocd.hp.arom.dco.aws.internal.cloud.***.***%2Fauth%2Fcallback&response_type=code&scope=openid+profile+email+groups&state=***"
argocd-server-79bc4c4f57-njhwx server time="2024-02-05T09:39:16Z" level=info msg="Callback: /auth/callback?code=***&state=***"
argocd-server-79bc4c4f57-njhwx server time="2024-02-05T09:39:17Z" level=info msg="Web login successful. Claims: {***}"

### dex server
time="2024-02-05T09:39:16Z" level=info msg="login successful: connector \"sesame\", username=\"***\", preferred_username=\"***\", email=\"***@*** (unverified)\", groups=[]"
```

**Comment by FredNass** [comment]
We are observing the same random behavior with argocd: v2.5.18+b56ef05

`msg="invalid client_secret on token request for client: argo-cd"` appears for each failed attempt in pod "argocd-dex-server" logs:

```
time="2024-02-21T15:13:54Z" level=info msg="username \"xxxxxxxxx\" mapped to entry uid=xxxxxxxx,ou=people,dc=xxxxxxxxx,dc=fr"
time="2024-02-21T15:13:54Z" level=info msg="performing ldap search ou=xxxxxxx,dc=xxxxxxxx,dc=fr sub (&(objectClass=groupOfNames)(member=uid=xxxxxxxx,ou=people,dc=xxxxxxxx,dc=fr))"
time="2024-02-21T15:13:54Z" level=info msg="login successful: connector \"ldap\", username=\"firstName lastName" \", preferred_username=\"\", email=\"xxxxxxxxx@xxxxxxxxx.fr\", groups=[\"xxxxxxxxxx\"]"
time="2024-02-21T15:13:54Z" level=info msg="invalid client_secret on token request for client: argo-cd"
```

Several attempts fail then suddenly one succeeds.

**Comment by rljohnsn** [comment]
I'm seeing similar experience on v2.9.0+9cf0c69

**Comment by elouanKeryell-Even** [comment]
A colleague of mine seems to have fixed it

He moved the OIDC client secret out of the helm chart by pre-creating the `argocd-secret` before `helm install`

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: argocd-secret
  namespace: argocd-system
  labels:
    app.kubernetes.io/name: argocd-secret
    app.kubernetes.io/part-of: argocd
type: Opaque
stringData:
  dex.config.CONNECTOR.clientSecret: *****
```

We don't encounter the error anymore. I don't understand why it fixes it tho

**Comment by ckav370** [comment]
+1 for this behaviour with Okta SAML and version v2.10.7

**Comment by husira** [comment]
we have noticed the same behaviour with v2.11.7. Randomly, it does not work and dex logs the error message:
`invalid client_secret on token request`

**Comment by abelhoula** [comment]
we have a similar issue with version 2.12. it does not work `invalid client_secret on token request for client: argocd"`

**Comment by garry-t** [comment]
Is it planed to be fixed ? Do we have work around for it?

**Comment by garry-t** [comment]
It was the nice 3.1 release, but what about this issue ? it is not possible to use SSO.

**Comment by CodeReaper** [comment]
@garry-t - I did a write-up of my investigation into this issue [here](https://codereaper.com/blog/2024/debugging-argo-cd-and-oidc-logins/).

Tldr; every time the secrets that Argo CD is dependent on changes, you will have restart the application to force it to load them.

Perhaps you (or others) can use this write-up to find workarounds and potentially share them here.

**Our workaround:**
Is to remove client secrets from ArgoCD-GitOps and have them be only GitOps. Meaning have them be deployed separately and ahead of ArgoCD in a deployment pipeline (and also to have the pipeline restart ArgoCD if those secrets had been changed).

**Comment by garry-t** [comment]
@CodeReaper thank you, really. I just created secret before argocd and it works. ! I will test it on a clean cluster, but I think it is really simple workaround
